sdk/webpubsub/azure-messaging-webpubsubclient/azure/messaging/webpubsubclient/_client.py

```diff
@@ -558,40 +693,50 @@ def _retry(self, func: Callable[[], None]):
                 )
                 time.sleep(delay_seconds)
 
-    def _call_back(self, callback_type: Union[CallbackType, str], *args):
+    def _call_back(self, callback_type: CallbackType, *args):
         for func in self._handler[callback_type]:
             # _call_back works in listener thread which must not be blocked so we have to execute the func
             # in new thread to avoid dead lock
             threading.Thread(target=func, args=args, daemon=True).start()
 
-    def _start_from_restarting(self):
+    def _start_from_restarting(self, reconnect_tried_times: Optional[int] = None):
         if self._state != WebPubSubClientState.DISCONNECTED:
             _LOGGER.warning("Client can be only restarted when it's Disconnected")
             return
 
         try:
-            self._start_core()
+            self._start_core(reconnect_tried_times)
         except Exception as e:
             self._state = WebPubSubClientState.DISCONNECTED
             raise e
 
-    def _handle_auto_reconnect(self):
+    def _handle_auto_reconnect(self, reconnect_tried_times: Optional[int] = None):
         _LOGGER.debug("start auto reconnect")
         success = False
-        attempt = 0
+        attempt = -1 if reconnect_tried_times is None else reconnect_tried_times
         while not self._is_stopping:
             try:
-                self._start_from_restarting()
+                attempt = attempt + 1
+                if attempt > 0:
+                    delay_seconds = self._reconnect_retry_policy.next_retry_delay(attempt)
+                    if not delay_seconds:
+                        break
+                    _LOGGER.debug(
+                        "Delay time for reconnect attempt %d: %ds",
+                        attempt,
+                        delay_seconds,
+                    )
+                    time.sleep(delay_seconds)
+                self._start_from_restarting(reconnect_tried_times)
+                if self._ws:
+                    self._ws.clear_reconnect_recover()
                 success = True
                 break
+            except ReconnectError:
```

Comment:
We used to keep retrying on Exception - but now we're breaking - is that going to change the customer experience?

Reply:
context: https://github.com/Azure/azure-sdk-for-python/pull/34764#discussion_r1568265617
`ReconnectError/RecoverError` will only raised in recover/reconnect logic and users will never feel them in running time.